### ISSUE TITLE
crash and freeze under Ubuntu

### ISSUE THREAD

**Issue by Simon818** [jmdaweb/NVDARemoteServer#5]
I'm running the remote server on several Ubuntu boxes with various versions. I have a while loop that's set to relaunch the remote server in debug mode if it crashes, but sometimes the server completely freezes. It displays errors when new connections are attempted, and pressing a ctrl+c does nothing. In fact, the only way to close it is to find the process name and use kill -9 to force it closed. This is very difficult for me to track down or even reproduce, I just know that it happens fairly regularly and causes interruptions to the server until I can get back in and fix it.

**Comment by Simon818** [comment]
Update: The server no longer crashes regularly, but still freezes quite a lot. It's the only serious issue I've seen. I have reports of it happening on Windows and Ubuntu, and can confirm that it happens under Ubuntu in several different places. I have a mac server at home which seems to remain unaffected, either because it has fewer people connecting to it or because of the OS change, I really couldn't say. Anything I can do to track this down? It's seriously debilitating to have to go in and restart this regularly, especially konwing that other people are being affected when it goes down.

**Comment by Simon818** [comment]
Also, it's worth adding that this seems to happen whether the server is running in debug mode, e.g. showing the server log on screen, or in daemon mode. I've tried both and had the same problem, and once connections start getting refused, the only way to kill it is with kill -9.

**Comment by jmdaweb** [comment]
Hello. Some days ago I had this problem on my centos 6 server. I have made some changes to print the full python trace to the console and the log file. I can't reproduce this isue, but if It happens again we will be able to know why it happens.

**Comment by Simon818** [comment]
I don't know if this will help, but here's what I get fairly often for tracebacks. This may also be unrelated, but it's all I'm seeing regularly.
Also, I know that this was a proper disconnect, not a disconnect initiated by the server.
Setting socket options...
Added a new client.
Client 16 joined channel XXXXXXXX
Socket error in client 16 while receiving data
Traceback (most recent call last):
  File "/home/simon/NVDARemoteServer/server.py", line 162, in handle_data
    data = self.buffer + self.socket.recv(8192)
  File "/usr/lib/python2.7/ssl.py", line 241, in recv
    return self.read(buflen)
  File "/usr/lib/python2.7/ssl.py", line 160, in read
    return self._sslobj.read(len)
error: [Errno 110] Connection timed out
Traceback (most recent call last):
  File "/home/simon/NVDARemoteServer/server.py", line 225, in close
    self.socket.shutdown(socket.SHUT_RDWR)
  File "/usr/lib/python2.7/ssl.py", line 292, in shutdown
    socket.shutdown(self, how)
  File "/usr/lib/python2.7/socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
error: [Errno 107] Transport endpoint is not connected
Client 16 has disconnected.
Client 16 removed.

**Comment by Simon818** [comment]
I don't know the exact cause, but I can tell you it happens consistently if a connection times out, rather than being closed. I discovered this completely by accident when my internet unexpectedly went down three times in a row. After it came back, the connection would time out after a few minutes, and the server would instantly become unresponsive. Even active connections would stop functioning and the script would have to be force-closed. I hope this helps. I would love to see this fixed sometime soon, as it's really the only thing stopping me from using this script everywhere.

**Comment by jmdaweb** [comment]
Hello.
Some days ago I experienced similar problems on my server. The program even refused accepting new connections! I had to kill it using NVDARemoteServer kill.
I have updated the script, now the clients dictionary use client ids instead of sockets as dictionary keys. Despite I have found in the log many broken connections, the server hasn't crashed. Can you test it on Ubuntu? If it works, I will close the isue.
Regards.

**Comment by Simon818** [comment]
I haven't had a crash yet, but I still get this. I believe this is an 
actual disconnect caused by the remote internet connection, but it still 
caused a traceback:
Client 2 joined channel ***
Socket error in client 2 while receiving data
Traceback (most recent call last):
   File "server.py", line 168, in handle_data
     data = self.buffer + self.socket.recv(16384)
   File "/usr/lib/python2.7/ssl.py", line 241, in recv
     return self.read(buflen)
   File "/usr/lib/python2.7/ssl.py", line 160, in read
     return self._sslobj.read(len)
error: [Errno 110] Connection timed out
Traceback (most recent call last):
   File "server.py", line 231, in close
     self.socket.shutdown(socket.SHUT_RDWR)
   File "/usr/lib/python2.7/ssl.py", line 292, in shutdown
     socket.shutdown(self, how)
   File "/usr/lib/python2.7/socket.py", line 224, in meth
     return getattr(self._sock,name)(*args)
error: [Errno 107] Transport endpoint is not connected
Client 2 has disconnected.
Client 2 removed.

**Comment by jmdaweb** [comment]
I have many similar errors on my server, but now the server doesn't crash, so I will close the isue. This is a client problem, related to the Internet connection.

**Comment by Simon818** [comment]
This crash does still happen. Additionally, I found this traceback when I looked in the terminal window before killing it:
Traceback (most recent call last):
  File "server.py", line 106, in accept_new_connection
    client_sock, addr = self.server_socket.accept()
  File "/usr/lib/python2.7/ssl.py", line 354, in accept
    suppress_ragged_eofs=self.suppress_ragged_eofs),
  File "/usr/lib/python2.7/ssl.py", line 143, in __init__
    self.do_handshake()
  File "/usr/lib/python2.7/ssl.py", line 305, in do_handshake
    self._sslobj.do_handshake()
SSLError: [Errno 8] _ssl.c:504: EOF occurred in violation of protocol

I'm not sure if this caused it, but it's definitely still a problem. The server lasted less than 24 hours.
